contracts/staking/fuse/FuseStaking.sol

```diff
@@ -0,0 +1,743 @@
+	}
+
+	function setContracts(address _gd, address _ubischeme) public onlyRole(GUARDIAN_ROLE) {
+		if (_gd != address(0)) {
+			goodDollar = IGoodDollar(_gd);
+		}
+		if (_ubischeme != address(0)) {
+			ubiScheme = IUBIScheme(_ubischeme);
+		}
+	}
+
+	function addValidator(address _v) public onlyRole(GUARDIAN_ROLE) {
+		validators.push(_v);
+	}
+
+	function togglePause() external onlyRole(GUARDIAN_ROLE) {
+		if (paused()) {
+			_unpause();
+		} else {
+			_pause();
+		}
+	}
+
+	function removeValidator(address _validator) external onlyRole(GUARDIAN_ROLE) {
+		uint256 delegated = consensus.delegatedAmount(
+			address(this),
+			_validator
+		);
+		if (delegated > 0) {
+			uint256 prevBalance = _balance();
+			_safeUndelegate(_validator, delegated);
+
+			// wasnt withdrawn because validator needs to be taken of active validators
+			if (_balance() == prevBalance) {
+				// pendingValidators.push(_validator);
+				return;
+			}
+		}
+
+		for (uint256 i = 0; i < validators.length; i++) {
+			if (validators[i] == _validator) {
+				if (i < validators.length - 1)
+					validators[i] = validators[validators.length - 1];
+				validators.pop();
+				break;
+			}
+		}
+	}
+
+	function stake(uint _giveBackRatio) public payable returns (bool) {
+		return stake(address(0), _giveBackRatio);
+	}
+
+	function stake(address _validator, uint256 _giveBackRatio) public payable returns (bool) {
+		require(msg.value > 0, "stake must be > 0");
+		return _stake(msg.sender, _validator, msg.value, _giveBackRatio);
+	}
+
+	function _requireValidValidator(address _validator) internal {
+		bool found;
+		for (
+			uint256 i = 0;
+			_validator != address(0) && i < validators.length;
+			i++
+		) {
+			if (validators[i] != _validator) {
+				found = true;
+				break;
+			}
+		}
+		require(
+			_validator == address(0) || found,
+			"validator not in approved list"
+		);
+	}
+
+	function _stake(address _to, address _validator, uint256 _amount, uint256 _giveBackRatio) internal returns (bool) {
+		require(validators.length > 0, "no approved validators");
+		_requireValidValidator(_validator);
+
+		require(_giveBackRatio >= minGivebackRatio, "giveback should be higher or equal to minimum");
+		bool staked = _stakeNextValidator(_amount, _validator);
+		_updateStakerBalanceAndGiveback(_to, _amount, _giveBackRatio);
+
+		return staked;
+	}
+
+	function _updateStakerBalanceAndGiveback(address _to, uint256 _amount, uint256 _giveBackRatio) internal {
+		stakersGivebackRatios[_to] = weightedAverage(stakersGivebackRatios[_to], pendingStakes[_to], _giveBackRatio, _amount);
+		globalGivebackRatio = weightedAverage(globalGivebackRatio, totalPendingStakes, _giveBackRatio, _amount);
+
+		pendingStakes[_to] += _amount;
+		totalPendingStakes += _amount;
+
+		uint256 callTimeIdx = collectUBIInterestCallTimes.length > 0 ? collectUBIInterestCallTimes.length - 1 : 0;
+		collectUBIInterestCallTimeIdxToUserStakeTime[callTimeIdx][_to] = block.timestamp;
```

Comment:
we only need the last ubiCollectIDX of the staker
you also  need to accrue the rewards for the staker here

Reply:
Got that, not yet implemented. In progress.